src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,561 @@
 #define SAMPLE_GAIN             (10000.0 / 32768.0)
 
 
-
-
-DEFINE_DEVICE_TYPE(HC55516, hc55516_device, "hc55516", "HC-55516")
-
-hc55516_device::hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
-	: hc55516_device(mconfig, HC55516, tag, owner, clock)
+//#####################################
+//                 COMMON
+//#####################################
+cvsd_device::cvsd_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_sound_interface(mconfig, *this)
+	, m_clock_state_push_cb(*this)
+	, m_digin_pull_cb(*this)
+	, m_digout_push_cb(*this)
+	, m_active_clock_edge(FALLING)
+	, m_shiftreg_mask(0)
+	, m_last_clock_state(false)
+	, m_buffered_bit(false)
+	, m_shiftreg(0)
+	, m_curr_sample(0)
+	, m_next_sample(0)
+	, m_samples_generated(0)
 {
 }
-
-hc55516_device::hc55516_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
-	: device_t(mconfig, type, tag, owner, clock),
-		device_sound_interface(mconfig, *this),
-		m_channel(nullptr),
-		m_active_clock_hi(0),
-		m_shiftreg_mask(0),
-		m_last_clock_state(0),
-		m_digit(0),
-		m_new_digit(0),
-		m_shiftreg(0),
-		m_curr_sample(0),
-		m_next_sample(0),
-		m_update_count(0),
-		m_filter(0),
-		m_integrator(0),
-		m_charge(0),
-		m_decay(0),
-		m_leak(0)
+//-------------------------------------------------
+//  start_common - common init and savestate reg for everything shared
+//-------------------------------------------------
+void cvsd_device::start_common(uint8_t shiftreg_mask, bool active_clock_edge)
 {
+	m_shiftreg_mask = shiftreg_mask;
+	m_active_clock_edge = active_clock_edge;
```

Comment:
This is an antipattern.  You should make `m_shiftreg_mask` and `m_active_clock_edge` members `const`-qualified and get the derived classes to supply values at construction time, then put the rest of the stuff here in `device_start`.  The derived classes should call the base `device_start` at the beginning of their `device_start`.

Reply:
Is there an example similar to what you want that I can follow elsewhere in the code? I don't understand exactly what I should be doing here.

Reply:
I got rid of start_common and changed the sub-devices to call the 'device_start' of their parent device in addition to handling their own save state registration/etc. I have not yet moved parameters into the constructor, and think it may be useful to leave at least some parameters non-constant and editable in the MAME debugger.

Reply:
This should now be resolved.